numba/rewrites.py

```diff
@@ -0,0 +1,42 @@
+from __future__ import print_function, division, absolute_import
+
+
+class Rewrite(object):
+    def match(self, block, typemap, calltypes):
+        return False
+
+    def apply(self):
+        raise NotImplementedError("Abstract Rewrite.apply() called!")
+
+
+class RewriteRegistry(object):
+    def __init__(self):
+        self.rewrites = []
+
+    def register(self, rewrite_cls):
+        assert issubclass(rewrite_cls, Rewrite)
+        self.rewrites.append(rewrite_cls)
+        return rewrite_cls
+
+    def apply(self, pipeline, blocks):
+        old_blocks = blocks.copy()
+        for rewrite_cls in self.rewrites:
+            # Exhaustively apply a rewrite until it stops matching.
+            rewrite = rewrite_cls(pipeline)
+            work_list = list(blocks.items())
+            while work_list:
+                key, block = work_list.pop()
+                matches = rewrite.match(block, pipeline.typemap,
```

Comment:
The match/apply split seems a bit artificial to me, in the sense that rewrite will need information already obtained by match (as seen as its sole implementation right now, where instance variables are used to hold that state). Is there any case where it would be useful to call "match" without a matching "apply" which follows immediately after the "if true" check?. This would also reduce a rewrite to be a single function. Note that a "match_and_apply" could notify of a "no rewrite" by returning None as the resulting block.